### ISSUE TITLE
cross cluster snapshot restore - s3 repository path error

### ISSUE THREAD

**Issue by SHolzhauer** [elastic/elasticsearch#59333]
**Elasticsearch version** (`bin/elasticsearch --version`): 7.8.0

**Description of the problem including expected versus actual behavior**:
Registering an S3 snapshot repository in two clusters, but in one with a trailing forward slash ("/") results in the snapshots
from the one without a trailing slash showing up in the second cluster but not being able to restore them.

**Steps to reproduce**:
 1. Register a snapshot repository on your first cluster
```bash
curl --location --request PUT 'https://127.0.0.1:9200/_snapshot/repo1' \
--header 'Content-Type: application/json' \
--data-raw '{
  "type": "s3",
  "settings": {
    "bucket": "s3-elk-stack",
    "base_path": "snapshots/7",
    "compress": true,
    "server_side_encryption": true,
    "protocol": "https",
    "readonly": false,
    "storage_class": "standard_ia"
  }
}
'
```
 2. Register the same snapshot repository on your second cluster but add a trailing "/" to the path
```bash
curl --location --request PUT 'https://127.0.0.1:9200/_snapshot/repo1' \
--header 'Content-Type: application/json' \
--data-raw '{
  "type": "s3",
  "settings": {
    "bucket": "s3-elk-stack",
    "base_path": "snapshots/7/",
    "compress": true,
    "server_side_encryption": true,
    "protocol": "https",
    "readonly": true,
    "storage_class": "standard_ia"
  }
}
'
```
 3. Make a snapshot on the first cluster
 4. Try to restore the snapshot you just made on the second cluster
    a. When restoring in Kibana, an error will be presented like "missing reponame:snapshot_name/xxx"
    b. When restoring using the API you will get more details amongst which you will notice it added a "/" to the complete path.

**Comment by original-brownbear** [comment]
Right, this is a real issue but very hard to solve cleanly. We actually have tests for this even I think, but since the slashes behave differently on real AWS S3 and Minio we didn't catch this apparently (on Minio they behave like actual path separators but on AWS S3 technically each `/` in a path is just another char like any other ... except for when we do the listing). There's a bit of an iffy question here of how to resolve this in a backwards compatible manner especially since we also support Minio.
I'll raise this during snapshot resiliency sync. One option is to deprecate base paths that end in a `/` in 7.x and simply strip trailing `/` in 8. That would align Minio and AWS S3 behavior fully and remove the weird ambiguity. The problem with this approach just is that it breaks repos on real S3 that were using a trailing `/` terminated path.

**Comment by SHolzhauer** [comment]
@original-brownbear I wasn't aware of Minio, one of the things I think that should be done at a minimum is document a warning/info in the docs? Atleast people can know this withouth accidently finding out.

On the road ahead, might I suggest a third step?
7.x deprecation
8.0 remove trailing `/` for new repo registration -> only do so in the registration process
8.x/9.0 remove trailing `/` everywhere.

**Comment by DaveCTurner** [comment]
We discussed this previously in https://github.com/elastic/elasticsearch/pull/46869 and I think that's still the right position: repos might exist for a very long time, and that the semantics of repeated `/` characters in S3 paths is implementation-defined, so we can't reliably fix the path configuration as proposed.

I think there's scope to improve _something_ given the OP's problem description, however. We should be consistent about whether the snapshots exist or not in the repo; IMO they shouldn't even exist if the base paths are not consistent.
